### PR TITLE
[FIX] hr_expense: bill partner child contact

### PR DESCRIPTION
**Pull request by agr-odoo**
Open a [DEMO] employee record
Set the Bank Account Number under 'Private Information'
Open the contact (res.partner) of [DEMO]
Set the parent contact to be the current company
Add a bank account on the contact for current company
Go to Expenses and create a new expense for [DEMO], with Paid By "Employee (to reimburse)"
Create Report, Submit Report, Approve, then Post Journal Entries.

Issue: On the created bill, partner will be the parent company and also
the bank account will be the one of the parent company.
When registering payment from:
- bill: the user will see the bank account of the company
- expense: the user will see the bank account of the employee

opw-3800134

### CODE REVIEW

**Comment by agr-odoo** [comment]
Hello @Levizar 
When it is possible could you give a look at this? I'm not sure it is functionally sound, but at the same time the current behavior of the system seems not consistent

**Comment by Levizar** [comment]
@JulienAlardot Could you have a look at it?

OFC, @agr-odoo is right, it doesn't make sense to reimburse the company instead of the employee.

File: addons/hr_expense/models/hr_expense.py [hunk 1]
```diff
@@ -1289,13 +1289,17 @@ def _prepare_payment_vals(self):
 
     def _prepare_bill_vals(self):
         self.ensure_one()
+        if self.payment_mode == 'company_account':
+            partner_id = self.employee_id.sudo().address_home_id.commercial_partner_id.id
```
Comment (JulienAlardot) [inline addons/hr_expense/models/hr_expense.py:1293]: This should never be the case, as only own_account expenses call this one

File: addons/hr_expense/models/hr_expense.py [hunk 2]
```diff
@@ -1289,13 +1289,17 @@ def _prepare_payment_vals(self):
...
+        else:
+            partner_id = self.employee_id.id
```
Comment (JulienAlardot) [inline addons/hr_expense/models/hr_expense.py:1295]: self.env['res.partner'] != self.env['hr.employee']

**Review by JulienAlardot — COMMENTED** [review]
100% a bug, but this fix confuses me a lot :thinking:

File: addons/hr_expense/models/hr_expense.py (hunk 2 above)
Comment (Levizar) [inline addons/hr_expense/models/hr_expense.py:1295]: Did you test the fix because I'm surprised it works.
We discussed it a bit with JUAL and unless I'm mistaken, `self.employee_id` isn't a `res.partner` and thus it is probably working by chance. :/

**Comment by agr-odoo** [comment]
My bad, it seems I left this `fix` in an inconsistent state and forgot about it :sweat_smile: 
However thanks for the functional check, I'll fix the fix and add a test for this case

**Comment by agr-odoo** [comment]
First of all I apologize If I pinged you on a *very much* draft PR 
I checked everything again and in my opinion this might be a technical limitation.

By default the move partner is the commercial partner, so in this use case, it would be the parent company, from which the system computes the appropriate bank account
If we force `partner_id` on the move vals, we can't set the `bank_partner_id` because it's a computed field and this will be used to get the `partner_bank_id` 
If we force `partner_id` and `partner_bank_id`, the bank account shown on the move is correct, but in the register payment wizard is not. The wizard `partner_bank_id` is not directly taken from the move but it is recomputed taking into account the partner of the payment term aml, which is, again, the commercial partner.

Then, if we look at the register payment wizardon the expense form, we find this override to select the employee account
https://github.com/odoo/odoo/blob/970e2d1c35b843f83f103124eaf0ecce97bda363/addons/hr_expense/wizard/account_payment_register.py#L18

In conclusion, this might be an expected limitation so the user should either:
- Register a payment from the expense payment widget
- Avoid setting a parent company for the employee related partner

**Comment by Levizar** [comment]
@agr-odoo Couldn´t we set a context key before retrieving the account to use and in the method that finaly retrieves it, check for the context key to early return the partnerś bank account and avoid using the company one?

**Comment by JulienAlardot** [comment]
Hello, after some discussion we decided not to fix this and add a warning about the case on expense & vendor bills asking to void that parent_id as it makes no sense to have it in the first place.
